icicle/appUtils/ntt/ntt.cuh

```diff
@@ -314,17 +355,47 @@ void ntt_inplace_batch_template(
                                                       // then max to allow more concurrent blocks on SM
   const int logn_shmem = is_shared_mem_enabled ? int(log(2 * num_threads) / log(2))
                                                : 0; // TODO: shared memory support only for types <= 32 bytes
-
   if (inverse) {
     if (is_shared_mem_enabled)
       ntt_template_kernel_shared<<<num_blocks, num_threads, shared_mem, stream>>>(
         d_inout, 1 << logn_shmem, d_twiddles, n, total_tasks, 0, logn_shmem);
+    cudaError_t cudaError = cudaGetLastError();
+    if(cudaError != cudaSuccess) {
+      printf("\nCUDA error: %s\n", cudaGetErrorString(cudaError));       
+      exit(1);
+    }
 
-    for (int s = logn_shmem; s < logn; s++) // TODO: this loop also can be unrolled
+    // E* d_inout_original = new E[n];
+    // for (int i = 0; i < n; i++) {
+    //   d_inout_original[i] = d_inout[i];
+    // }
+    for (int s = logn_shmem; s < logn - 1; s++) // TODO: this loop also can be unrolled
     {
       ntt_template_kernel<E, S>
-        <<<num_blocks, num_threads, 0, stream>>>(d_inout, n, d_twiddles, n, total_tasks, s, false);
+        <<<num_blocks / 2, num_threads / 2, 0, stream>>>(d_inout, n / 2, d_twiddles, n / 2, total_tasks / 2, s, false);
+      ntt_template_kernel<E, S>
+        <<<num_blocks / 2, num_threads / 2, 0, stream>>>(d_inout + n / 2, n / 2, d_twiddles + n / 2, n / 2, total_tasks / 2, s, false);
+      cudaError_t cudaError = cudaGetLastError();
+      if(cudaError != cudaSuccess) {
+        printf("\nCUDA error: %s\n", cudaGetErrorString(cudaError));       
+        exit(1);
+      }
     }
+    ntt_template_kernel<E, S>
```

Comment:
I think this line poses an issue for correctness in some cases: consider NTTs of size less than `1 << logn_shmem` - they are computed fully inside the `ntt_template_kernel_shared` kernel. However, this line introduces one extra call that breaks correctness, this can be checked by verifying that for sizes of 1 << 10 or less ntt and intt fail to cancel each other but if this call and it's analogue in the forward NTT are commented out, they start to cancel